File: app/src/main/cpp/filter/base_filter.cpp
//
// Created by templechen on 2019/3/15.
//

#include "base_filter.h"
#include "../common/matrix_util.h"
#include <android/surface_texture.h>
#include <android/surface_texture_jni.h>
#include <android/asset_manager_jni.h>

base_filter::base_filter(int textureId) {
    oesTextureId = static_cast<GLuint>(textureId);
    initMatrix();
}

base_filter::~base_filter() {
    auto textures = new GLuint[1];
    textures[0] = oesTextureId;
    glDeleteTextures(1, textures);

    glDeleteShader(vertexShader);
    glDeleteShader(fragmentShader);
    glDeleteProgram(program);

    delete vertex_shader_string;
    delete fragment_shader_string;

    delete textureMatrix;
    delete coordMatrix;
}

void base_filter::initProgram() {
    readShaderString();
    vertexShader = loadShader(GL_VERTEX_SHADER, vertex_shader_string);
    fragmentShader = loadShader(GL_FRAGMENT_SHADER, fragment_shader_string);
    program = createShaderProgram(vertexShader, fragmentShader);
}

void base_filter::doFrame() {
    glUseProgram(program);

    glActiveTexture(GL_TEXTURE0);
    glBindTexture(GL_TEXTURE_EXTERNAL_OES, oesTextureId);

    GLint vertexCount = sizeof(myVertex) / (sizeof(myVertex[0]) * 2);
    aPositionLocation = glGetAttribLocation(program, aPosition);
    glEnableVertexAttribArray(aPositionLocation);
    glVertexAttribPointer(aPositionLocation, 2, GL_FLOAT, GL_FALSE, 8, myVertex);

    aTextureCoordinateLocation = glGetAttribLocation(program, aTextureCoordinate);
    glEnableVertexAttribArray(aTextureCoordinateLocation);
    glVertexAttribPointer(aTextureCoordinateLocation, 2, GL_FLOAT, GL_FALSE, 8, texture);

    uTextureMatrixLocation = glGetUniformLocation(program, uTextureMatrix);
//    glUniformMatrix4fv(uTextureMatrixLocation, 1, GL_FALSE, this->transformMatrix);
    glUniformMatrix4fv(uTextureMatrixLocation, 1, GL_FALSE, this->textureMatrix->m);

    uCoordMatrixLocation = glGetUniformLocation(program, uCoordMatrix);
    glUniformMatrix4fv(uCoordMatrixLocation, 1, GL_FALSE, this->coordMatrix->m);

    glDrawArrays(GL_TRIANGLES, 0, vertexCount);
    glDisableVertexAttribArray(aPositionLocation);
    glDisableVertexAttribArray(aTextureCoordinateLocation);
    glBindTexture(GL_TEXTURE_EXTERNAL_OES, 0);

}

void base_filter::readShaderString() {
    vertex_shader_string = {
            "uniform mat4 uTextureMatrix;\n"
            "uniform mat4 uCoordMatrix;\n"
            "attribute vec4 aPosition;\n"
            "attribute vec4 aTextureCoordinate;\n"
            "varying vec2 vTextureCoord;\n"
            "void main()\n"
            "{\n"
            "    vTextureCoord = (uTextureMatrix * aTextureCoordinate).xy;\n"
            "    gl_Position = uCoordMatrix * aPosition;\n"
            "}\n"
    };
    fragment_shader_string = {
            "#extension GL_OES_EGL_image_external : require\n"
            "precision mediump float;\n"
            "uniform samplerExternalOES uTextureSampler;\n"
            "varying vec2 vTextureCoord;\n"
            "void main()\n"
            "{\n"
            "    gl_FragColor = texture2D(uTextureSampler, vTextureCoord);\n"
            "}\n"
    };
}

void base_filter::initMatrix() {
    textureMatrix = new ESMatrix();
    textureMatrix->m[0] = 1.0f;
    textureMatrix->m[1] = 0.0f;
    textureMatrix->m[2] = 0.0f;
    textureMatrix->m[3] = 0.0f;
    textureMatrix->m[4] = 0.0f;
    textureMatrix->m[5] = -1.0f;
    textureMatrix->m[6] = 0.0f;
    textureMatrix->m[7] = 0.0f;
    textureMatrix->m[8] = 0.0f;
    textureMatrix->m[9] = 0.0f;
    textureMatrix->m[10] = 1.0f;
    textureMatrix->m[11] = 0.0f;
    textureMatrix->m[12] = 0.0f;
    textureMatrix->m[13] = 1.0f;
    textureMatrix->m[14] = 0.0f;
    textureMatrix->m[15] = 1.0f;

    coordMatrix = new ESMatrix();
    setIdentityM(coordMatrix);
}

void
base_filter::setScaleAndTransform(float scaleX, float scaleY, float transformX, float transformY,
                                  int rotateDegrees) {
    setIdentityM(coordMatrix);
    scaleM(coordMatrix, 0, scaleX, scaleY, 1.0f);
    translateM(coordMatrix, 0, transformX / scaleX, transformY / scaleY, 0.0f);
//    rotateM(coordMatrix, rotateDegrees, 0.0f, 0.0f, 1.0f);
    updateVertexCoord(rotateDegrees);
}

void base_filter::setVideoSize(int width, int height) {
    this->videoWidth = width;
    this->videoHeight = height;
}

void base_filter::setViewSize(int width, int height) {
    this->viewWidth = width;
    this->viewHeight = height;
}

void base_filter::updateVertexCoord(float degrees) {
    if (viewWidth > 0 && viewHeight > 0 && videoWidth > 0 && videoHeight > 0) {
        float ratio = videoWidth * 1.0f / videoHeight;
        float radian = -degrees * PI / 180;

        PointF *pointF1 = new PointF();
        pointF1->x = 1.0;
        pointF1->y = 1.0;
        PointF *pointF2 = new PointF();
        pointF2->x = -1.0;
        pointF2->y = 1.0;
        PointF *pointF3 = new PointF();
        pointF3->x = -1.0;
        pointF3->y = -1.0;
        PointF *pointF4 = new PointF();
        pointF4->x = 1.0;
        pointF4->y = -1.0;
        PointF *vertexCoord[4] = {
                pointF1, pointF2, pointF3, pointF4
        };

        for (int i = 0; i < 4; ++i) {
            PointF *pointF = vertexCoord[i];
            PointF *newPointF = new PointF();
            newPointF->x = (pointF->x) * cos(radian) - (pointF->y / ratio) * sin(radian);
            newPointF->y = ((pointF->x) * sin(radian) + (pointF->y / ratio) * cos(radian)) * ratio;
            vertexCoord[i] = newPointF;
        }
        myVertex[0] = vertexCoord[0]->x;
        myVertex[1] = vertexCoord[0]->y;
        myVertex[2] = vertexCoord[1]->x;
        myVertex[3] = vertexCoord[1]->y;
        myVertex[4] = vertexCoord[2]->x;
        myVertex[5] = vertexCoord[2]->y;
        myVertex[6] = vertexCoord[0]->x;
        myVertex[7] = vertexCoord[0]->y;
        myVertex[8] = vertexCoord[2]->x;
        myVertex[9] = vertexCoord[2]->y;
        myVertex[10] = vertexCoord[3]->x;
        myVertex[11] = vertexCoord[3]->y;
    };

}

File: app/src/main/java/com/example/templechen/videoshaderdemo/filter/BaseFilter.kt
package com.example.templechen.videoshaderdemo.filter

import android.content.Context
import android.opengl.GLES11Ext
import android.opengl.GLES20
import android.opengl.GLES30
import android.opengl.Matrix
import android.util.Log
import com.example.templechen.videoshaderdemo.GLUtils
import com.example.templechen.videoshaderdemo.R
import java.nio.FloatBuffer

open class BaseFilter {

    companion object {
        const val aPosition = "aPosition"
        const val uTextureMatrix = "uTextureMatrix"
        const val aTextureCoordinate = "aTextureCoordinate"
        const val uTextureSampler = "uTextureSampler"
        const val uMVPMatrix = "uMVPMatrix"
    }

    protected var context: Context
    protected var mOESTextureId: Int
    protected var floatBuffer: FloatBuffer
    protected var vertexShader: Int = -1
    protected var fragmentShader: Int = -1
    protected var program: Int = -1
    var transformMatrix = floatArrayOf(
        1.0f, 0.0f, 0.0f, 0.0f,
        0.0f, -1.0f, 0.0f, 0.0f,
        0.0f, 0.0f, 1.0f, 0.0f,
        0.0f, 1.0f, 0.0f, 1.0f
    )
    var baseMVPMatrix = FloatArray(16) {0f}

    protected var aPositionLocation = -1
    protected var uTextureMatrixLocation = -1
    protected var aTextureCoordinateLocation = -1
    protected var uTextureSamplerLocation = -1
    protected var uMVPMatrixLocation = -1

    constructor(context: Context, oesTextureId: Int) {
        this.context = context
        mOESTextureId = oesTextureId
        floatBuffer = GLUtils.createBuffer(GLUtils.vertexData)
        Matrix.setIdentityM(baseMVPMatrix, 0)
    }

    open fun initProgram() {
        vertexShader = GLUtils.loadShader(
            GLES30.GL_VERTEX_SHADER,
            GLUtils.readShaderFromResource(context, R.raw.base_vertex_shader)
        )
        fragmentShader = GLUtils.loadShader(
            GLES30.GL_FRAGMENT_SHADER,
            GLUtils.readShaderFromResource(context, R.raw.base_fragment_shader)
        )
        program = GLUtils.createProgram(vertexShader, fragmentShader)
    }

    open fun drawFrame() {
        GLES30.glUseProgram(program)

        aPositionLocation = GLES30.glGetAttribLocation(program, aPosition)
        aTextureCoordinateLocation = GLES30.glGetAttribLocation(program, aTextureCoordinate)
        uTextureMatrixLocation = GLES30.glGetUniformLocation(program, uTextureMatrix)
        uTextureSamplerLocation = GLES30.glGetUniformLocation(program, uTextureSampler)
        uMVPMatrixLocation = GLES30.glGetUniformLocation(program, uMVPMatrix)

        GLES30.glActiveTexture(GLES30.GL_TEXTURE0)
        GLES30.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, mOESTextureId)

        GLES30.glUniform1i(uTextureSamplerLocation, 0)
        GLES30.glUniformMatrix4fv(uTextureMatrixLocation, 1, false, transformMatrix, 0)
        GLES30.glUniformMatrix4fv(uMVPMatrixLocation, 1, false, baseMVPMatrix, 0)

        floatBuffer.position(0)
        GLES30.glEnableVertexAttribArray(aPositionLocation)
        GLES30.glVertexAttribPointer(aPositionLocation, 2, GLES30.GL_FLOAT, false, 16, floatBuffer)
        floatBuffer.position(2)
        GLES30.glEnableVertexAttribArray(aTextureCoordinateLocation)
        GLES30.glVertexAttribPointer(aTextureCoordinateLocation, 2, GLES30.GL_FLOAT, false, 16, floatBuffer)

        GLES30.glDrawArrays(GLES30.GL_TRIANGLES, 0, 6)

        GLES30.glDisableVertexAttribArray(aPositionLocation)
        GLES30.glDisableVertexAttribArray(aTextureCoordinateLocation)
        GLES30.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, 0)
    }

    open fun drawOverlayFrame() {
        GLES30.glUseProgram(program)

        aPositionLocation = GLES30.glGetAttribLocation(program, aPosition)
        aTextureCoordinateLocation = GLES30.glGetAttribLocation(program, aTextureCoordinate)
        uTextureMatrixLocation = GLES30.glGetUniformLocation(program, uTextureMatrix)
        uTextureSamplerLocation = GLES30.glGetUniformLocation(program, uTextureSampler)
        uMVPMatrixLocation = GLES30.glGetUniformLocation(program, uMVPMatrix)

        GLES30.glActiveTexture(GLES30.GL_TEXTURE1)
        GLES30.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, mOESTextureId)

        GLES30.glUniform1i(uTextureSamplerLocation, 0)
        GLES30.glUniformMatrix4fv(uTextureMatrixLocation, 1, false, transformMatrix, 0)
        GLES30.glUniformMatrix4fv(uMVPMatrixLocation, 1, false, baseMVPMatrix, 0)

        floatBuffer.position(0)
        GLES30.glEnableVertexAttribArray(aPositionLocation)
        GLES30.glVertexAttribPointer(aPositionLocation, 2, GLES30.GL_FLOAT, false, 16, floatBuffer)
        floatBuffer.position(2)
        GLES30.glEnableVertexAttribArray(aTextureCoordinateLocation)
        GLES30.glVertexAttribPointer(aTextureCoordinateLocation, 2, GLES30.GL_FLOAT, false, 16, floatBuffer)

        GLES30.glDrawArrays(GLES30.GL_TRIANGLES, 0, 6)

        GLES30.glDisableVertexAttribArray(aPositionLocation)
        GLES30.glDisableVertexAttribArray(aTextureCoordinateLocation)
        GLES30.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, 0)
    }

    fun setMVPMatrixScaleAndTransform(scaleX: Float, scaleY: Float, scrollX: Float, scrollY: Float) {
        Matrix.setIdentityM(baseMVPMatrix, 0)
        Matrix.scaleM(baseMVPMatrix, 0, scaleX, scaleY, 1.0f)
        Matrix.translateM(baseMVPMatrix, 0, scrollX, scrollY, 1.0f)
    }

    open fun release() {
        GLES30.glDeleteProgram(program)
        program = 0
        GLES30.glDeleteShader(vertexShader)
        vertexShader = 0
        GLES30.glDeleteShader(fragmentShader)
        fragmentShader = 0
        floatBuffer.clear()
    }

}